sn_node/src/node/messaging/agreement.rs

```diff
@@ -107,6 +108,36 @@ impl Node {
             }
         }
 
+        // In case there is a node already having the same connection info,
+        // we need to:
+        //   1, if the existing node is valid, reject the incoming agreement.
+        //   2, if the existing node is invalid, accept the incoming agreement
+        //      AND propose an offline against the existing node.
+        //
+        // The existing node is valid when:
+        //   1, Its age is lower than the incoming node by more than 1.
+        //   2, Its age is higher than the incoming node by just 1.
+        if let Some(existing_node) = self
+            .network_knowledge
+            .get_name_by_conn_info(&new_info.addr())
+        {
+            let existing_age = existing_node[XOR_NAME_LEN - 1];
+            let incoming_age = new_info.age();
```

Comment:
Consider modifying the `get_name_by_conn_info()` helper to return a `Peer` or a `NodeState` so that we can use the `age()` methods.

```suggestion
        if let Some(existing_peer) = self
            .network_knowledge
            .get_peer_by_conn_info(&new_info.addr())
        {
            let existing_age = existing_peer.age();
            let incoming_age = new_info.age();
```